crates/uv/tests/it/python_install.rs

```
@@ -982,9 +982,87 @@ fn python_install_preview_broken_link() {
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    Installed Python 3.12.8 in [TIME]
+     + cpython-3.12.8-[PLATFORM]
+    "###);
+
+    // But prefer explicit requests
+    uv_snapshot!(context.filters(), context.python_install().arg("3.11").env(EnvVars::UV_PYTHON, "3.12"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    Installed Python 3.11.11 in [TIME]
+     + cpython-3.11.11-[PLATFORM]
+    "###);
+
+    // Same with uninstall
+    uv_snapshot!(context.filters(), context.python_uninstall().env(EnvVars::UV_PYTHON, "3.12"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    Searching for Python versions matching: Python 3.12
+    Uninstalled Python 3.12.8 in [TIME]
+     - cpython-3.12.8-[PLATFORM]
+    "###);
+
+    uv_snapshot!(context.filters(), context.python_uninstall().arg("3.11").env(EnvVars::UV_PYTHON, "3.12"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    Searching for Python versions matching: Python 3.11
+    Uninstalled Python 3.11.11 in [TIME]
+     - cpython-3.11.11-[PLATFORM]
+    "###);
+
+    // Install again so we can test interaction with `uninstall --all`
+    uv_snapshot!(context.filters(), context.python_install().arg("3.12"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    Installed Python 3.12.8 in [TIME]
+     + cpython-3.12.8-[PLATFORM]
+    "###);
+
+    // We should ignore `UV_PYTHON` here
+    uv_snapshot!(context.filters(), context.python_uninstall().arg("--all").env(EnvVars::UV_PYTHON, "3.11"), @r###"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    error: the argument '--all' cannot be used with '<TARGETS>...'
+
+    Usage: uv python uninstall --all --install-dir <INSTALL_DIR> <TARGETS>...
+
+    For more information, try '--help'.
+    "###);
```

This test failure is problematic — this needs to work still.

---

I see you mentioned this at

>  As the --all arg conflicts with target it also conflicts UV_PYTHON env variable

We might need to implement this differently to work around this. Perhaps `--all` should just override the targets?

---

I do agree that `-all` should override targets as I don't see a case where a user would add -all argument without intending to delete all the version. So `all` should override targets set through argument or environment. Do we need to give the user a warning (I personally don't think it's needed) like:

> Setting targets with -all is redundant and just result in all managed python versions being uninstalled

---

Yeah I don't think a warning is needed, it's a weird case but it seems fine to just ignore the target silently unless someone reports a confusing interaction.

---

Ah are you suggesting that we should remove the conflict entirely and let someone do `uv python uninstall foo --all`? I think that could be too confusing :/

---

yes that is what I had in mind that if `--all` is set target shouldn't matter . I do agree invalid targets with `--all` not giving an error can be a bit confusing. Currently `uv python uninstall foo -all` doesn't really tell the user that their target is invalid it just tells them of the conflict of target with `--all`

So what I understand is you suggest that `--all` should override target but if it target is invalid it should throw an error

---

Oh sorry I mean `uv python uninstall 3.11 --all` would be confusing too — just used `foo` as a silly example.

---

oh so you suggest the target set as an argument should behave differently in case of use with `--all` (throw an error citing a conflict) vs target set through the env variable should be ignored silently. Apologies that took me a minute to understand :/

That makes sense for usage here the only concern I have is are there any other env variable that behave like this in uv. Most of the env variable I have seen they are sort of a replacement for just having to type to argument again and again and behave exactly the same as the cli argument. So this behaving differently might be not be very idiomatic. I also might just be overthinking this one.

---

> oh so you suggest the target set as an argument should behave differently in case of use with --all (throw an error citing a conflict) vs target set through the env variable should be ignored silently. Apologies that took me a minute to understand :/

Yeah, that's the suggestion. No problem on the confusion.

> Most of the env variable I have seen they are sort of a replacement for just having to type to argument again and again and behave exactly the same as the cli argument. So this behaving differently might be not be very idiomatic. I also might just be overthinking this one.

I think it's okay in this case. Though I'm not sure of the implications implementation-wise.

---

I couldn't find a suitable way to implement  it using clap's derive API. My workaround was writing a custom validation check that is ran after the parsing and before the uninstall Settings are resolved. The implementation does get a bit ugly in order to replicate the format of the error message without using the derive API